File: 541.ReverseStringII/Reverse.java
class Solution {
    public String reverseStr(String s, int k) {
        if (s == null) {
            return null;
        }

        int end = s.length() / (2*k) * (2*k);

        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < end; i += 2*k) {
            sb.append(reverse(s.substring(i, i+k))).append(s.substring(i+k, i+2*k));
        }

        int begin = end;
        end = s.length() - begin > k ? begin + k : s.length();

        sb.append(reverse(s.substring(begin, end)));

        if (end < s.length()) {
            sb.append(s.substring(end, s.length()));
        }

        return sb.toString();
    }

    public String reverse(String s) {
        char[] s_char = s.toCharArray();
        StringBuffer sb = new StringBuffer();
        for (int i = s_char.length-1; i >= 0; i--) {
            sb.append(s_char[i]);
        }
        return sb.toString();
    }
}

File: 541.ReverseStringII/reverse.cc
class Solution {
public:
    string reverseStr(string s, int k) {
        int end = s.length() / (2*k) * (2*k);
        for (int i = 0; i < end; i += 2*k) {
            reverse(s, i, i+k);
        }
        return reverse(s, end, int(s.length()) - end > k ? end + k : int(s.length()));
    }

    string reverse(string& s, int begin, int end) {
        for (int i = begin; i < (end - begin) / 2 + begin; i++) {
            char tmp = s[i];
            s[i] = s[end - (i - begin) - 1];
            s[end - (i - begin) - 1] = tmp;
        }
        return s;
    }
};
